cmake/HPX_AddModule.cmake

```diff
@@ -0,0 +1,166 @@
+# Copyright (c) 2019 Auriane Reverdell
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+function(add_hpx_module name)
+  # Retrieve arguments
+  set(options DEPRECATION_WARNINGS)
+  set(one_value_args COMPATIBILITY_HEADERS)
+  set(multi_value_args SOURCES HEADERS COMPAT_HEADERS DEPENDENCIES CMAKE_SUBDIRS)
+  cmake_parse_arguments(${name} "${options}" "${one_value_args}" "${multi_value_args}" ${ARGN})
+
+  # Compatibility header off is not specified
+  if (NOT ${name}_COMPATIBILITY_HEADERS)
+      set(${name}_COMPATIBILITY_HEADERS OFF)
+  endif()
+
+  # Main directories of the module
+  set(SOURCE_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/src")
+  hpx_debug("Add module ${name}: SOURCE_ROOT: ${SOURCE_ROOT}")
+  set(HEADER_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/include")
+  hpx_debug("Add module ${name}: HEADER_ROOT: ${HEADER_ROOT}")
+  set(COMPAT_HEADER_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility")
+  hpx_debug("Add module ${name}: COMPAT_HEADER_ROOT: ${COMPAT_HEADER_ROOT}")
+
+  string(TOUPPER ${name} name_upper)
+
+  # HPX options
+
+  hpx_option(HPX_${name_upper}_WITH_TESTS
+    BOOL
+    "Build HPX ${name} module tests. (default: ON)"
+    ON ADVANCED
+    CATEGORY "Modules")
+
+  if(${name}_DEPRECATION_WARNINGS)
+    hpx_option(HPX_${name_upper}_WITH_DEPRECATION_WARNINGS
+      BOOL
+      "Enable warnings for deprecated facilities. (default: ${HPX_WITH_DEPRECATION_WARNINGS})"
+      ${HPX_WITH_DEPRECATION_WARNINGS} ADVANCED
+      CATEGORY "Modules")
+    if(HPX_${name}_WITH_DEPRECATION_WARNINGS)
+      hpx_add_config_define_namespace(
+          DEFINE HPX_${name_upper}_HAVE_DEPRECATION_WARNINGS
+          NAMESPACE ${name_upper})
+    endif()
+  endif()
+  
+  if(${name}_COMPATIBILITY_HEADERS)
+    # Added in 1.4.0
+    hpx_option(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS
+      BOOL
+      "Enable compatibility headers for old headers"
+      ON ADVANCED
+      CATEGORY "Modules")
+    if(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS)
+      hpx_add_config_define_namespace(
+          DEFINE HPX_${name_upper}_HAVE_COMPATIBILITY_HEADERS
+          NAMESPACE ${name_upper})
+    endif()
+  endif()
+
+  # Write full path for the sources files
+  include(HPX_CMakeUtils)
+  prepend(sources ${SOURCE_ROOT} ${${name}_SOURCES})
+  prepend(headers ${HEADER_ROOT} ${${name}_HEADERS})
```

Comment:
Will this do the right thing even if the files are already fully specified?

Reply:
No, but should we specify the full path here ? cause the structure in each module is the same : `src/`, `include/`, `include_compatibility`

Reply:
I think we should be able to go with relative paths for now. It forces us to use the same structure in all the modules. If we need absolute paths at some point in the future I think we can easily add an option for that when needed.

Reply:
Fair point.